internal/extender/resource.go

```diff
@@ -418,6 +422,101 @@ func (s *SparkSchedulerExtender) getNodes(ctx context.Context, nodeNames []strin
+	executorSparkLabel, ok := executor.Labels[common.SparkAppIDLabel]
+	if !ok {
+		return "", false, werror.ErrorWithContextParams(ctx, "Executor does not have a Spark app id label, could not create label selector")
+	}
+	applicationPods, err := s.getSparkApplicationPodsForExecutor(ctx, executor, executorSparkLabel)
+	if err != nil {
+		return "", false, err
+	}
+	scheduledPods := filterToRunningPods(ctx, applicationPods)
+	azs, err := filterToAZs(ctx, scheduledPods, s)
+	if err != nil {
+		return "", false, err
+	}
+	svc1log.FromContext(ctx).Info("Pods are scheduled in zones", svc1log.SafeParam("zones", azs.ToSlice()))
+	if azs.Size() > 1 {
+		return "", false, nil
+	}
+	if azs.Size() == 0 {
+		return "", false, werror.ErrorWithContextParams(ctx, "Application has no scheduled pods, can't make scheduling decisions based on AZ")
+	}
+	return azs.ToSlice()[0], true, nil
+}
+
+func filterToAZs(ctx context.Context, runningPods []*v1.Pod, s *SparkSchedulerExtender) (utils.StringSet, error) {
+	azs := utils.NewStringSet(len(runningPods))
+	// Get all node AZs where pods are running
+	for _, pod := range runningPods {
+		nodeName := pod.Spec.NodeName
+		node, err := s.nodeLister.Get(nodeName)
+		if err != nil {
+			return nil, err
+		}
+		zoneLabel, ok := node.Labels[v1.LabelTopologyZone]
+		if !ok {
+			return nil, werror.ErrorWithContextParams(ctx, "Could not read zone label from node, unable to make scheduling decisions based on AZ")
+		}
+		azs.Add(zoneLabel)
+	}
+	return azs, nil
+}
+
+func filterToRunningPods(ctx context.Context, applicationPods []*v1.Pod) []*v1.Pod {
+	scheduledPods := make([]*v1.Pod, 0, len(applicationPods))
+	for _, pod := range applicationPods {
+		if pod.Status.Phase == v1.PodRunning {
+			scheduledPods = append(scheduledPods, pod)
+		}
+	}
+	logRunningPods(ctx, scheduledPods)
+	return scheduledPods
+}
+
+func logRunningPods(ctx context.Context, scheduledPods []*v1.Pod) {
+	scheduledPodNames := make([]string, 0, len(scheduledPods))
+	for _, pod := range scheduledPods {
+		scheduledPodNames = append(scheduledPodNames, pod.Name)
+	}
+	svc1log.FromContext(ctx).Info("Filtered to running application pods", svc1log.SafeParam("scheduledApplicationPods", scheduledPodNames))
+}
+
+func (s *SparkSchedulerExtender) getSparkApplicationPodsForExecutor(ctx context.Context, executor *v1.Pod, executorSparkLabel string) ([]*v1.Pod, error) {
+	applicationPods, err := s.podLister.Pods(executor.Namespace).List(labels.Set(map[string]string{common.SparkAppIDLabel: executorSparkLabel}).AsSelector())
+	if err != nil {
+		return nil, err
+	}
+	applicationPodNames := make([]string, 0, len(applicationPods))
+	for _, pod := range applicationPods {
+		applicationPodNames = append(applicationPodNames, pod.Name)
+	}
+	svc1log.FromContext(ctx).Info("Found existing application pods", svc1log.SafeParam("applicationPodNames", applicationPodNames))
```

Comment:
nit: can we put this, `logRunningPods`, and other logging into a separate file? It makes functions longer and harder to read.